libraries/aws/provisioning/src/aws_iot_provisioning_api.c

```diff
@@ -392,6 +401,37 @@ AwsIotProvisioningError_t AwsIotProvisioning_Init( uint32_t mqttTimeoutMs )
 
 /*-----------------------------------------------------------*/
 
+AwsIotProvisioningError_t _timedWaitForServerResponse( uint32_t timeoutMs )
+{
+    AwsIotProvisioningError_t status = AWS_IOT_PROVISIONING_SUCCESS;
+
+    /* Wait for response from the server. */
+    if( IotSemaphore_TimedWait( &_activeOperation.responseReceivedSem,
+                                timeoutMs ) == false )
+    {
+        status = AWS_IOT_PROVISIONING_TIMEOUT;
+    }
+
+    /* There can be an edge case of receiving the server response right after the timeout has occurred.
+     * If such an edge case has occurred and causes the subscription callback to finish its execution after we timeout,
+     * we use the outcome of the callback execution instead of treating this case as a "timeout".
+     * Rationale - This API is not intended to have an "exact" implementation of server timeout, as the completion of the
+     * the operation is more valuable for the customer than having an exact behavior of server timeout. */
+    IotMutex_Lock( &_activeOperation.lock );
```

Comment:
This does not seem right, it seems like it is just reducing the probability of the race condition and not handling it in a reliable way? What happens if you take the lock, check the status, the response has not arrived, you release the lock and THEN the response arrives?

The only way in a case like this to reliably handle the race condition is to change state inside the lock (not just the local variable "status", but the state which will cause the response to be dropped if it arrives after this check instead of normally processing it.

This is probably better done by taking the lock inside the if statement above, checking the active operation inside that lock and then updating the local status AND the active operation status and then releasing the lock.

Reply:
Discussed on Chime, and Cobus provided the suggestion to remove the race-condition handling logic as it does not fully resolve the race condition problem. 
Thus, we decided to remove the logic that attempts to partially solve the race condition case. The cost of not solving the race-condition is (just) a re-attempt of the request to the service API.